src/coreComponents/constitutive/fluid/twophasefluid/TwoPhaseFluid.cpp

```diff
@@ -0,0 +1,299 @@
+
+
+  /* ???????
+     // call to correctly set member array tertiary sizes on the 'main' material object
+     resizeFields( 0, 0 );
+
+     // set labels on array wrappers for plottable fields
+     setLabels();
+   */
+}
+
+
+void
+TwoPhaseFluid::readTable( string const & fileName,
+                          integer minRowLength,
+                          array1d< array1d< real64 > > & data )
+{
+  std::ifstream is( fileName );
+  GEOS_ERROR_IF( !is.is_open(),
+                 "TwoPhaseFluid: could not open file: " << fileName );
+
+  // Read line-by-line until eof
+  string str;
+  while( std::getline( is, str ) )
+  {
+    // Remove whitespace and end-of-line characters, if any
+    str = stringutilities::trim( str, " \r" );
+
+    // Remove # and -- (Eclipse-style) comments
+    str = stringutilities::removeStringAndFollowingContent( str, "#" );
+    str = stringutilities::removeStringAndFollowingContent( str, "--" );
+
+    // Skip empty or comment-only strings
+    if( str.empty() )
+      continue;
+
+    // Add and read a new line entry
+    array1d< real64 > newLine = stringutilities::fromStringToArray< real64 >( str );
+    if( !newLine.empty() )
+    {
+      data.emplace_back( std::move( newLine ) );
+    }
+  }
+
+  is.close();
+
+  for( localIndex i = 0; i < data.size(); ++i )
+  {
+    GEOS_ERROR_IF( data[i].size() < minRowLength,
+                   "TwoPhaseFluid: too few entries in row " << i << " of table " << fileName
+                                                            << ", minimum " << std::to_string( minRowLength ) << " required" );
+  }
+}
+
+
+void TwoPhaseFluid::fillData( integer const ip,
+                              array1d< array1d< real64 > > const & tableValues )
+{
+  array1d< array1d< real64 > > pressureCoords( 1 );
+  pressureCoords[0].resize( tableValues.size() );
+  array1d< real64 > density( tableValues.size() );
+  array1d< real64 > viscosity( tableValues.size() );
+
+  for( localIndex i = 0; i < tableValues.size(); ++i )
+  {
+    GEOS_THROW_IF_NE_MSG( tableValues[i].size(), 3,
+                          getFullName() << ": three columns (pressure, density, and viscosity) are expected",
+                          InputError );
+
+    pressureCoords[0][i] = tableValues[i][0];
+    density[i] = tableValues[i][1];
+    viscosity[i] = tableValues[i][2];
+  }
+
+  string const densityTableName = getName() + "DensityPhase" + GEOS_FMT( "{}", ip );
+  string const viscosityTableName = getName() +  "ViscosityPhase" + GEOS_FMT( "{}", ip );
+  m_densityTableNames.emplace_back( densityTableName );
+  m_viscosityTableNames.emplace_back( viscosityTableName );
+
+  FunctionManager & functionManager = FunctionManager::getInstance();
+
+  TableFunction & tableDensity =
+    dynamicCast< TableFunction & >( *functionManager.createChild( "TableFunction", densityTableName ) );
+  tableDensity.setTableCoordinates( pressureCoords, { units::Pressure } );
```

Comment:
Do we need to check that `pressureCoords` is increasing? Or perhaps the `TableFunction` object already checks this?

Reply:
Yes, the `TableFunction` has already a built-in check to ensure coordinates are strictly increasing